021-server-defined-attributes.md

```diff
@@ -0,0 +1,120 @@
+- POST `/v1/databases/:databaseId/collections/:collectionId/attributes/server`
+- DELETE `/v1/databases/:databaseId/collections/:collectionId/attributes/server/:key`
+
+We can allow users to create indexes for new attributes from the `createIndex` endpoint.
+
+We need to add a check in `createDocument` and `updateDocument` endpoints to populate the values of these attributes if they exist for client-side requests.
+
+###  Workers / Commands
+<!-- Do we need new workers or commands for this feature? List and describe them and their API signatures -->
+No new workers are anticipated.
+
+###  Supporting Libraries
+<!-- Do we need new libraries for this feature? Mention which, define the file structure, and different interfaces -->
+No new changes required in libraries.
+
+### Data Structures
+<!-- Do we need new data structures for this feature? Describe and explain the new collections and attributes -->
+#### In appwrite/appwrite
+The new attributes would add attributes to `attributes` collection in `$projectCollections` for each optional attribute.
+
+```
+$projectCollections = array_merge([
+  'attributes' => [
+        '$collection' => ID::custom(Database::METADATA),
+        '$id' => ID::custom('attributes'),
+        'name' => 'Attributes',
+        'attributes' => [
+            [
+                '$id' => ID::custom('createdBy'),
+                'type' => Database::VAR_STRING,
+                'format' => '',
+                'size' => 256,
+                'signed' => true,
+                'required' => false,
+                'default' => null,
+                'array' => false,
+                'filters' => [],
+            ],
+...
+]
+```
+
+### SDKs
+<!-- Do we need to update our SDKs for this feature? Describe how -->
+SDKs will be updated to accommodate the changes in the API endpoints.
+
+### Breaking Changes
+<!-- Will this feature introduce any breaking changes? How can we achieve backward compatability -->
+No breaking changes are anticipated as this is an additive feature.
+
+### Documentation & Content
+<!-- What documentation do we need to update or add for this feature? -->
+The documentation will be updated to inform developers about the new endpoints and how to add these optional attributes in their collections.
+
+## Reliability
+
+### Security
+<!-- How will we secure this feature? -->
+
+### Scaling
+<!-- How will we scale this feature? -->
+This feature will rely on existing infrastructure, so additional scaling measures are not anticipated.
+
+### Benchmarks
+<!-- How will we benchmark this feature? -->
+No new benchmarks are required as existing infrastructure will support the feature.
+
+### Tests (UI, Unit, E2E)
+<!-- How will we test this feature? -->
+- End-to-end testing for each new endpoint and functionality.
+- User Interface testing if a GUI component is added.
+
+## Open Questions
+<!-- List of things we need to figure out or farther discuss -->
+N/A
```

Comment:
Which ones should we enable by default? None?

Reply:
None by default. We add them when users create them.